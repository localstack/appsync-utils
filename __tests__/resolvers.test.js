/* test full resolver pipelines 
Note: for the request-context the naming must be `arguments` when passing it. 
Within the request it can be resolved to both, e.g. `ctx.arguments` and `ctx.args`
*/

import { checkResolverValid } from "./helpers";
import { util } from "..";
import * as fs from 'fs/promises';
import * as path from 'path';
import { fileURLToPath } from 'url';

async function importCodeFromFile(stub) {
  const root = path.dirname(fileURLToPath(import.meta.url));
  const fullPath = path.join(path.resolve(root), stub);
  return await fs.readFile(fullPath, { encoding: "utf8" });
}

describe("dynamodb resolvers", () => {
  test("something", async () => {
    const code = `
    export function request(ctx) {
        return {
            operation: "Query",
            index: "nameIndex",
            select : "ALL_ATTRIBUTES",
            query: {
                "expression" : "#name = :name",
                "expressionNames": {
                    "#name": "name",
                },
                "expressionValues" : {
                    ":name" : util.dynamodb.toDynamoDB(ctx.arguments.filter.line),
                }
            },
            filter: {
                "expression" : "#shift = :shift",
                "expressionNames": {
                    "#shift": "shift",
                },
                "expressionValues" : {
                    ":shift" : util.dynamodb.toDynamoDB(ctx.arguments.filter.shift),
                }
            },
        };
    }

    export function response(ctx) {
        return ctx.result.items;
    }
    `;

    const requestContext = {
      arguments: {
        filter: {
          line: "test",
          shift: 10,
        },
      },
    };

    await checkResolverValid(code, requestContext, "request");

    const responseContext = {
      result: {
        items: [
          { a: 10 },
        ],
      },
    };

    await checkResolverValid(code, responseContext, "response");
  });
});

describe("rds resolvers", () => {
  describe("toJsonObject import paths", () => {
    const context = {
      result: JSON.stringify({
        sqlStatementResults: [
          {
            numberOfRecordsUpdated: 0,
            records: [
              [
                {
                  stringValue: "Mark Twain",
                },
                {
                  stringValue: "Adventures of Huckleberry Finn",
                },
                {
                  stringValue: "978-1948132817",
                },
              ],
              [
                {
                  stringValue: "Jack London",
                },
                {
                  stringValue: "The Call of the Wild",
                },
                {
                  stringValue: "978-1948132275",
                },
              ],
            ],
            columnMetadata: [
              {
                isSigned: false,
                isCurrency: false,
                label: "author",
                precision: 200,
                typeName: "VARCHAR",
                scale: 0,
                isAutoIncrement: false,
                isCaseSensitive: false,
                schemaName: "",
                tableName: "Books",
                type: 12,
                nullable: 0,
                arrayBaseColumnType: 0,
                name: "author",
              },
              {
                isSigned: false,
                isCurrency: false,
                label: "title",
                precision: 200,
                typeName: "VARCHAR",
                scale: 0,
                isAutoIncrement: false,
                isCaseSensitive: false,
                schemaName: "",
                tableName: "Books",
                type: 12,
                nullable: 0,
                arrayBaseColumnType: 0,
                name: "title",
              },
              {
                isSigned: false,
                isCurrency: false,
                label: "ISBN-13",
                precision: 15,
                typeName: "VARCHAR",
                scale: 0,
                isAutoIncrement: false,
                isCaseSensitive: false,
                schemaName: "",
                tableName: "Books",
                type: 12,
                nullable: 0,
                arrayBaseColumnType: 0,
                name: "ISBN-13",
              },
            ],
          },
        ],
      }),
    };

    test("default", async () => {
      let importStatement;
      if (process.env.TEST_TARGET === "AWS_CLOUD") {
        importStatement =
          "import { toJsonObject } from '@aws-appsync/utils/rds';";
      } else {
        importStatement = "import { toJsonObject } from '../rds';";
      }
      const code =
        importStatement +
        "\n" +
        `
      export function request(ctx) {
          return toJsonObject(ctx.result);
      }

      export function response(ctx) {
      }
    `;
      await checkResolverValid(code, context, "request");
    });

    test("fully qualified", async () => {
      const code = `
      export function request(ctx) {
          return util.rds.toJsonObject(ctx.result);
      }

      export function response(ctx) {
      }
    `;
      await checkResolverValid(code, context, "request");
    });
  });

  describe("typehints", () => {
    test("UUID", async () => {
      const code = `
            export function request(ctx) {
                return rds.typeHint.UUID(ctx.args.id);
            }

            export function response(ctx) {
            }
            `;

      const context = {
        arguments: {
          id: "abc123",
        },
      };

      await checkResolverValid(code, context, "request");
    });

    test("TIMESTAMP", async () => {
      const code = `
            export function request(ctx) {
                return rds.typeHint.TIMESTAMP(ctx.args.id);
            }

            export function response(ctx) {
            }
            `;

      const context = {
        arguments: {
          id: new Date(2023, 1, 1),
        },
      };

      await checkResolverValid(code, context, "request");
    });
  });

  // https://docs.aws.amazon.com/appsync/latest/devguide/resolver-reference-rds-js.html
  test("toJsonObject", async () => {
    const responseContext = {
      "result": JSON.stringify({
        "sqlStatementResults": [
          {
            "numberOfRecordsUpdated": 0,
            "records": [
              [
                {
                  "stringValue": "Mark Twain"
                },
                {
                  "stringValue": "Adventures of Huckleberry Finn"
                },
                {
                  "stringValue": "978-1948132817"
                }
              ],
              [
                {
                  "stringValue": "Jack London"
                },
                {
                  "stringValue": "The Call of the Wild"
                },
                {
                  "stringValue": "978-1948132275"
                }
              ]
            ],
            "columnMetadata": [
              {
                "isSigned": false,
                "isCurrency": false,
                "label": "author",
                "precision": 200,
                "typeName": "VARCHAR",
                "scale": 0,
                "isAutoIncrement": false,
                "isCaseSensitive": false,
                "schemaName": "",
                "tableName": "Books",
                "type": 12,
                "nullable": 0,
                "arrayBaseColumnType": 0,
                "name": "author"
              },
              {
                "isSigned": false,
                "isCurrency": false,
                "label": "title",
                "precision": 200,
                "typeName": "VARCHAR",
                "scale": 0,
                "isAutoIncrement": false,
                "isCaseSensitive": false,
                "schemaName": "",
                "tableName": "Books",
                "type": 12,
                "nullable": 0,
                "arrayBaseColumnType": 0,
                "name": "title"
              },
              {
                "isSigned": false,
                "isCurrency": false,
                "label": "ISBN-13",
                "precision": 15,
                "typeName": "VARCHAR",
                "scale": 0,
                "isAutoIncrement": false,
                "isCaseSensitive": false,
                "schemaName": "",
                "tableName": "Books",
                "type": 12,
                "nullable": 0,
                "arrayBaseColumnType": 0,
                "name": "ISBN-13"
              }
            ]
          }
        ]
      }),
    };

    const code = `
        export function request(ctx) {}

        export function response(ctx) {
            return rds.toJsonObject(ctx.result);
        }
        `;

    await checkResolverValid(code, responseContext, "response");
  });
  test("toJsonObject insert or update", async () => {
    const responseContext = {
      "result": JSON.stringify({
        "sqlStatementResults": [{
          "numberOfRecordsUpdated": 1,
          "generatedFields": []
        }]
      })
    };

    const code = `
        export function request(ctx) {}

        export function response(ctx) {
            return rds.toJsonObject(ctx.result);
        }
        `;

    await checkResolverValid(code, responseContext, "response");
  })

  test("toJsonObject update and select", async () => {
    const responseContext = {
      "result": JSON.stringify({
        "sqlStatementResults": [{
          "numberOfRecordsUpdated": 1,
          "generatedFields": []
        },
        {
          "numberOfRecordsUpdated": 0,
          "records": [
            [
              {
                "stringValue": "Mark Twain"
              },
              {
                "stringValue": "Adventures of Huckleberry Finn"
              },
              {
                "stringValue": "978-1948132817"
              }
            ],
            [
              {
                "stringValue": "Jack London"
              },
              {
                "stringValue": "The Call of the Wild"
              },
              {
                "stringValue": "978-1948132275"
              }
            ]
          ],
          "columnMetadata": [
            {
              "isSigned": false,
              "isCurrency": false,
              "label": "author",
              "precision": 200,
              "typeName": "VARCHAR",
              "scale": 0,
              "isAutoIncrement": false,
              "isCaseSensitive": false,
              "schemaName": "",
              "tableName": "Books",
              "type": 12,
              "nullable": 0,
              "arrayBaseColumnType": 0,
              "name": "author"
            },
            {
              "isSigned": false,
              "isCurrency": false,
              "label": "title",
              "precision": 200,
              "typeName": "VARCHAR",
              "scale": 0,
              "isAutoIncrement": false,
              "isCaseSensitive": false,
              "schemaName": "",
              "tableName": "Books",
              "type": 12,
              "nullable": 0,
              "arrayBaseColumnType": 0,
              "name": "title"
            },
            {
              "isSigned": false,
              "isCurrency": false,
              "label": "ISBN-13",
              "precision": 15,
              "typeName": "VARCHAR",
              "scale": 0,
              "isAutoIncrement": false,
              "isCaseSensitive": false,
              "schemaName": "",
              "tableName": "Books",
              "type": 12,
              "nullable": 0,
              "arrayBaseColumnType": 0,
              "name": "ISBN-13"
            }
          ]
        }
        ]
      })
    };

    const code = `
        export function request(ctx) {}

        export function response(ctx) {
            return rds.toJsonObject(ctx.result);
        }
        `;

    await checkResolverValid(code, responseContext, "response");
  })


  describe("mysql", () => {
    test("raw string", async () => {
      const code = `
      export function request(ctx) {
        const { id } = ctx.args;
        const updateQuery = \`UPDATE "document" set "data" = 10 WHERE id = \${id}\`;
        return rds.createMySQLStatement(updateQuery);
      }

      export function response(ctx) {}
    `;

      const context = {
        arguments: {
          id: "adb626eb-4ce5-452a-a917-3943a37f202b",
        },
      };

      await checkResolverValid(code, context, "request");
    });

    test("sql tagged template", async () => {
      const code = `
      export function request(ctx) {
        const { id, text } = ctx.args;
        const s1 = rds.sql\`insert into Post(id, text) values(\${rds.typeHint.UUID(id)}, \${text})\`;
        const s2 = rds.sql\`select * from Post where id = \${id}\`;
        return rds.createMySQLStatement(s1, s2);
      }

      export function response(ctx) {}
    `;

      const context = {
        arguments: {
          id: "adb626eb-4ce5-452a-a917-3943a37f202b",
          text: "hello world",
        },
      };

      await checkResolverValid(code, context, "request");
    });

    test("type hints", async () => {
      const code = `
      export function request(ctx) {
        const whereClause = { and:[
          { id: { eq: rds.typeHint.UUID(ctx.args.id) } },
          { started: { lt: rds.typeHint.TIMESTAMP(ctx.args.started) } } 
        ] }; 
        return rds.createMySQLStatement(rds.select({
          table: "UserGroup",
          where: whereClause,
          }));
      }

      export function response(ctx) {}
      `
      const requestContext = {
        arguments: {
          id: "1232",
          name: "hello",
          started: new Date(2022, 2, 2),
        }
      };

      await checkResolverValid(code, requestContext, "request");

    });

    test("select", async () => {
      const code = `
    export function request(ctx) {
        const whereClause = { or: [
          { name: { eq: 'Stephane'} },
          { id: { gt: 10 } }
      ]}
        return rds.createMySQLStatement(rds.select({
            table: "UserGroup",
            where: whereClause,
            limit: 10,
            offset: 1,
            columns: ['id', 'name'],
            orderBy: [{column: 'name'}, {column: 'id', dir: 'DESC'}]
        }));
    }

    export function response(ctx) {}
    `;

      const requestContext = {};

      await checkResolverValid(code, requestContext, "request");

    });

    test("update", async () => {
      const code = `
        export function request(ctx) {
            const { input: { id, ...values }, condition } = ctx.args;
            const where = {
                ...condition,
                id: { eq: id },
            };
            const updateStatement = rds.update({
                table: 'persons',
                values,
                where,
            });

            return rds.createMySQLStatement(updateStatement)
        }
        export function response(ctx) {}
    `;
      const requestContext = {
        arguments: {
          input: {
            id: "abc123",
            name: "name",
            birthday: "today",
            country: "home",
          },
        },
      };

      await checkResolverValid(code, requestContext, "request");
    });

    test("insert", async () => {
      const code = `
        export function request(ctx) {
        const { input: values } = ctx.args;
        const insertStatement = rds.insert({ table: 'persons', values });
        
        return rds.createMySQLStatement(insertStatement)
        }

        export function response(ctx) {}
        `;

      const requestContext = {
        arguments: {
          input: {
            name: "test",
          },
        }
      };

      await checkResolverValid(code, requestContext, "request");
    });

    test("remove", async () => {
      const code = `
      export function request(ctx) {
          const id = ctx.args.id;
          const where = { id: { eq: id } };
          const deleteStatement = rds.remove({
              table: 'persons',
              where: where,
          });
      
          return rds.createMySQLStatement(deleteStatement);
        }
      export function response(ctx) {}
  `;

      const requestContext = {
        arguments: {
          id: "1232"
        }
      };

      await checkResolverValid(code, requestContext, "request");

    });
  });

  describe("postgresql", () => {
    test("raw string", async () => {
      const code = `
      export function request(ctx) {
        const { id } = ctx.args;
        const updateQuery = \`UPDATE "document" set "data" = 10 WHERE id = \${id}\`;
        return rds.createPgStatement(updateQuery);
      }

      export function response(ctx) {}
    `;

      const context = {
        arguments: {
          id: "adb626eb-4ce5-452a-a917-3943a37f202b",
        },
      };

      await checkResolverValid(code, context, "request");
    });

    test("json functions in raw string", async () => {
      const code = await importCodeFromFile("./functions/pgRawQueryJson.js");

      const context = {
        arguments: {
          id: "adb626eb-4ce5-452a-a917-3943a37f202b",
          key: "key",
          value: "value",
        },
      };

      await checkResolverValid(code, context, "request");
    });

    test("sql tagged template", async () => {
      const code = `
      export function request(ctx) {
        const { id, text } = ctx.args;
        const s1 = rds.sql\`insert into Post(id, text) values(\${rds.typeHint.UUID(id)}, \${text})\`;
        const s2 = rds.sql\`select * from Post where id = \${id}\`;
        return rds.createPgStatement(s1, s2);
      }

      export function response(ctx) {}
    `;

      const context = {
        arguments: {
          id: "adb626eb-4ce5-452a-a917-3943a37f202b",
          text: "hello world",
        },
      };

      await checkResolverValid(code, context, "request");
    });

    test("type hints", async () => {
      const code = `
      export function request(ctx) {
        const whereClause = { and:[
          { id: { eq: rds.typeHint.UUID(ctx.args.id) } },
          { started: { lt: rds.typeHint.TIMESTAMP(ctx.args.started) } } 
        ] }; 
        return rds.createPgStatement(rds.select({
          table: "UserGroup",
          where: whereClause,
          }));
      }

      export function response(ctx) {}
      `
      const requestContext = {
        arguments: {
          id: "1232",
          name: "hello",
          started: new Date(2022, 2, 2),
        }
      };

      await checkResolverValid(code, requestContext, "request");

    });

    test("select", async () => {
      const code = `
    export function request(ctx) {
        const whereClause = { or: [
          { name: { eq: 'Stephane'} },
          { id: { gt: 10 } }
      ]}
        return rds.createPgStatement(rds.select({
            table: "UserGroup",
            where: whereClause,
            limit: 10,
            offset: 1,
            columns: ['id', 'name'],
            orderBy: [{column: 'name'}, {column: 'id', dir: 'DESC'}]
        }));
    }

    export function response(ctx) {}
    `;

      const requestContext = {};

      await checkResolverValid(code, requestContext, "request");

    });

    test("update", async () => {
      const code = `
        export function request(ctx) {
            const { input: { id, ...values }, condition } = ctx.args;
            const where = {
                ...condition,
                id: { eq: id },
            };
            const updateStatement = rds.update({
                table: 'persons',
                values,
                where,
            });

            return rds.createPgStatement(updateStatement)
        }
        export function response(ctx) {}
    `;
      const requestContext = {
        arguments: {
          input: {
            id: "abc123",
            name: "name",
            birthday: "today",
            country: "home",
          },
        },
      };

      await checkResolverValid(code, requestContext, "request");
    });


    test("insert", async () => {
      const code = `
        export function request(ctx) {
        const { input: values } = ctx.args;
        const insertStatement = rds.insert({ table: 'persons', values, returning: "*" });
        
        return rds.createPgStatement(insertStatement)
        }

        export function response(ctx) {}
        `;

      const requestContext = {
        arguments: {
          input: {
            name: "test",
          },
        }
      };

      await checkResolverValid(code, requestContext, "request");
    });

    test("remove", async () => {
      const code = `
      export function request(ctx) {
          const id = ctx.args.id;
          const where = { id: { eq: id } };
          const deleteStatement = rds.remove({
              table: 'persons',
              where: where,
              returning: ['id', 'name'],
          });
      
          return rds.createPgStatement(deleteStatement);
        }
      export function response(ctx) {}
  `;

      const requestContext = {
        arguments: {
          id: "1232"
        }
      };

      await checkResolverValid(code, requestContext, "request");

    });
  });
});

