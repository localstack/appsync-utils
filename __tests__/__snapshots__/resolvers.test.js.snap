// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamodb resolvers something 1`] = `
{
  "filter": {
    "expression": "#shift = :shift",
    "expressionNames": {
      "#shift": "shift",
    },
    "expressionValues": {
      ":shift": {
        "N": 10,
      },
    },
  },
  "index": "nameIndex",
  "operation": "Query",
  "query": {
    "expression": "#name = :name",
    "expressionNames": {
      "#name": "name",
    },
    "expressionValues": {
      ":name": {
        "S": "test",
      },
    },
  },
  "select": "ALL_ATTRIBUTES",
}
`;

exports[`dynamodb resolvers something 2`] = `
[
  {
    "a": 10,
  },
]
`;

exports[`rds resolvers mysql insert 1`] = `
{
  "statements": [
    "INSERT INTO \`persons\` (\`name\`) VALUES (:P0)",
  ],
  "variableMap": {
    ":P0": "test",
  },
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers mysql raw string 1`] = `
{
  "statements": [
    "UPDATE "document" set "data" = 10 WHERE id = adb626eb-4ce5-452a-a917-3943a37f202b",
  ],
  "variableMap": {},
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers mysql remove 1`] = `
{
  "statements": [
    "DELETE FROM \`persons\` WHERE \`id\` = :P0",
  ],
  "variableMap": {
    ":P0": "1232",
  },
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers mysql select 1`] = `
{
  "statements": [
    "SELECT \`id\`, \`name\` FROM \`UserGroup\` WHERE (\`name\` = :P0) OR (\`id\` > :P1) ORDER BY \`name\` ASC, \`id\` DESC LIMIT :P2 OFFSET :P3",
  ],
  "variableMap": {
    ":P0": "Stephane",
    ":P1": 10,
    ":P2": 10,
    ":P3": 1,
  },
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers mysql sql tagged template 1`] = `
{
  "statements": [
    "insert into Post(id, text) values(:P0, :P1)",
    "select * from Post where id = :P2",
  ],
  "variableMap": {
    ":P0": "adb626eb-4ce5-452a-a917-3943a37f202b",
    ":P1": "hello world",
    ":P2": "adb626eb-4ce5-452a-a917-3943a37f202b",
  },
  "variableTypeHintMap": {
    ":P0": "UUID",
  },
}
`;

exports[`rds resolvers mysql type hints 1`] = `
{
  "statements": [
    "SELECT * FROM \`UserGroup\` WHERE (\`id\` = :P0) AND (\`started\` < :P1)",
  ],
  "variableMap": {
    ":P0": "1232",
    ":P1": "2022-03-02T00:00:00.000Z",
  },
  "variableTypeHintMap": {
    ":P0": "UUID",
    ":P1": "TIMESTAMP",
  },
}
`;

exports[`rds resolvers mysql update 1`] = `
{
  "statements": [
    "UPDATE \`persons\` SET \`name\` = :P0, \`birthday\` = :P1, \`country\` = :P2 WHERE \`id\` = :P3",
  ],
  "variableMap": {
    ":P0": "name",
    ":P1": "today",
    ":P2": "home",
    ":P3": "abc123",
  },
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers postgresql insert 1`] = `
{
  "statements": [
    "INSERT INTO "persons" ("name") VALUES (:P0) RETURNING *",
  ],
  "variableMap": {
    ":P0": "test",
  },
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers postgresql json functions in raw string 1`] = `
{
  "statements": [
    "UPDATE "document" set "data" = jsonb_set("data", '{key}', '"value" WHERE id = adb626eb-4ce5-452a-a917-3943a37f202b",
  ],
  "variableMap": {},
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers postgresql raw string 1`] = `
{
  "statements": [
    "UPDATE "document" set "data" = 10 WHERE id = adb626eb-4ce5-452a-a917-3943a37f202b",
  ],
  "variableMap": {},
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers postgresql remove 1`] = `
{
  "statements": [
    "DELETE FROM "persons" WHERE "id" = :P0 RETURNING "id", "name"",
  ],
  "variableMap": {
    ":P0": "1232",
  },
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers postgresql select 1`] = `
{
  "statements": [
    "SELECT "id", "name" FROM "UserGroup" WHERE ("name" = :P0) OR ("id" > :P1) ORDER BY "name" ASC, "id" DESC LIMIT :P2 OFFSET :P3",
  ],
  "variableMap": {
    ":P0": "Stephane",
    ":P1": 10,
    ":P2": 10,
    ":P3": 1,
  },
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers postgresql sql tagged template 1`] = `
{
  "statements": [
    "insert into Post(id, text) values(:P0, :P1)",
    "select * from Post where id = :P2",
  ],
  "variableMap": {
    ":P0": "adb626eb-4ce5-452a-a917-3943a37f202b",
    ":P1": "hello world",
    ":P2": "adb626eb-4ce5-452a-a917-3943a37f202b",
  },
  "variableTypeHintMap": {
    ":P0": "UUID",
  },
}
`;

exports[`rds resolvers postgresql type hints 1`] = `
{
  "statements": [
    "SELECT * FROM "UserGroup" WHERE ("id" = :P0) AND ("started" < :P1)",
  ],
  "variableMap": {
    ":P0": "1232",
    ":P1": "2022-03-02T00:00:00.000Z",
  },
  "variableTypeHintMap": {
    ":P0": "UUID",
    ":P1": "TIMESTAMP",
  },
}
`;

exports[`rds resolvers postgresql update 1`] = `
{
  "statements": [
    "UPDATE "persons" SET "name" = :P0, "birthday" = :P1, "country" = :P2 WHERE "id" = :P3",
  ],
  "variableMap": {
    ":P0": "name",
    ":P1": "today",
    ":P2": "home",
    ":P3": "abc123",
  },
  "variableTypeHintMap": {},
}
`;

exports[`rds resolvers toJsonObject 1`] = `
[
  [
    {
      "ISBN-13": "978-1948132817",
      "author": "Mark Twain",
      "title": "Adventures of Huckleberry Finn",
    },
    {
      "ISBN-13": "978-1948132275",
      "author": "Jack London",
      "title": "The Call of the Wild",
    },
  ],
]
`;

exports[`rds resolvers toJsonObject import paths default 1`] = `
[
  [
    {
      "ISBN-13": "978-1948132817",
      "author": "Mark Twain",
      "title": "Adventures of Huckleberry Finn",
    },
    {
      "ISBN-13": "978-1948132275",
      "author": "Jack London",
      "title": "The Call of the Wild",
    },
  ],
]
`;

exports[`rds resolvers toJsonObject import paths fully qualified 1`] = `
[
  [
    {
      "ISBN-13": "978-1948132817",
      "author": "Mark Twain",
      "title": "Adventures of Huckleberry Finn",
    },
    {
      "ISBN-13": "978-1948132275",
      "author": "Jack London",
      "title": "The Call of the Wild",
    },
  ],
]
`;

exports[`rds resolvers toJsonObject insert or update 1`] = `
[
  [
    {},
  ],
]
`;

exports[`rds resolvers toJsonObject update and select 1`] = `
[
  [
    {},
  ],
  [
    {
      "ISBN-13": "978-1948132817",
      "author": "Mark Twain",
      "title": "Adventures of Huckleberry Finn",
    },
    {
      "ISBN-13": "978-1948132275",
      "author": "Jack London",
      "title": "The Call of the Wild",
    },
  ],
]
`;

exports[`rds resolvers typehints TIMESTAMP 1`] = `
{
  "type": "TIMESTAMP",
  "value": "2023-02-01T00:00:00.000Z",
}
`;

exports[`rds resolvers typehints UUID 1`] = `
{
  "type": "UUID",
  "value": "abc123",
}
`;
